### ISSUE TITLE
With Quantum, is there better chance for compatibility?

### ISSUE THREAD

**Issue by punassuming** [joaomgcd/JoinFirefox#4]
With the shift to Firefox Quantum and the change in the backend to support webextensions, is there better chance to get this working? Looking at the other issue, it seems like it is tied up with GCM compatibility.

**Comment by joaomgcd** [comment]
Unfortunately there's one big issue to add Firefox support: it doesn't support push messaging in extensions. I've filed a request about it here: https://bugzilla.mozilla.org/show_bug.cgi?id=1378096 Hopefully they can add it eventually

**Comment by Camology** [comment]
How do things like pushbullet work on firefox then?

**Comment by joaomgcd** [comment]
They use a very expensive solution (which is probably why they need a subscription) which I can't afford unfortunately...

**Comment by deostroll** [comment]
I am subscribing to this thread with the hope I will get notified of progress. Ready to help if needed...

**Comment by mrandreastoth** [comment]
Subscribing for the same reason. Unable to offer my help other than for testing, unfortunately.

By the way, is there any other way to go about solving this until Firefox support is implemented such as running an external server on the same device to handle the push messaging and act as a bridge somehow?

**Comment by joaomgcd** [comment]
Unfortunately that's not an option, sorry!

**Comment by Camology** [comment]
So the only solution is an expensive solution as you say, would this be a one time licensing cost or an ongoing service cost? How much would we need to crowdfund to get this operation up and running ?

**Comment by joaomgcd** [comment]
It's an ongoing cost. Unfortunately unless I start charging a subscription this wouldn't be possible...

**Comment by mrandreastoth** [comment]
I'm trying to understand what the options are and what costs are associated with each of these. Knowing this information someone may then be able to help (for example, say you needed a server, knowing this someone may then offer to share their server for free with you).

By the way, I'm curious, how come the other options don't invoke any costs (i.e., the Chrome extension and the Windows app)?

If the Windows 10 module can do what it can do and do it for free, could you then not write a Windows 10 module that operated like the current app but instead of raising notifications, etc., it instead spoke to the Firefox extension? In fact, I know of a few Firefox extensions that make use of such helper modules in order to solve Firefox and Chrome limitations. The extension Send To Chrome/Firefox uses such an external module, for example (I may have the extension name wrong).

What I'm struggling to understand is, if your current Windows 10 app can implement all the necessary functionality without invoking any cost to you, then why would the suggested bridging module invoke any costs?

**Comment by joaomgcd** [comment]
The reason those other services don't have any costs is because Google and Microsoft have free push messaging services in Chrome and Windows :)

Chrome uses [FCM](https://firebase.google.com/docs/cloud-messaging/) and windows uses [Windows Push Notification Services](https://docs.microsoft.com/en-us/windows/uwp/design/shell/tiles-and-notifications/windows-push-notification-services--wns--overview).

Unfortunately I don't know of a way of making a windows app talk to a Firefox extension...

**Comment by mrandreastoth** [comment]
The following is for native Windows apps and Firefox communication. Not sure if it would work for UWP apps.

https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Native_messaging

**Comment by joaomgcd** [comment]
Thank you. It's still a messy solution, to have to install the UWP app for it to work, but I'll try figuring out if it's possible. Thanks

**Comment by mrandreastoth** [comment]
As a developer to another, I fully understand you wish to do it right, but when no other viable option exists, then it's an acceptable workaround that I know people will understand and accept. Then, once Firefox is fixed, you just have to update the Firefox extension and revoke the workaround UWP app from the store and everything will continue to operate the right way.

**Comment by joaomgcd** [comment]
Thanks for understanding :)

**Comment by ase1590** [comment]
would it be a better solution to instead wrap the whole thing in an Electron application, that way it can run independently of a browser and use the OS's notification system?

**Comment by deostroll** [comment]
Not quite. That becomes a desktop app. It is a different UX altogether
which is not relevant for browsing. What is important to have here is the
ability to share URLs of what you're browsing, copying stuff to and from
clipboard, etc.

**Comment by Rayquaza01** [comment]
I don't know if GCM is available in electron apps, but if it is, it might be possible to tie it into the browser with native messaging. So the UI would be in the browser/extension, but the pushes are actually being received by the desktop app and are just relayed to the browser.

**Comment by vertigo220** [comment]
> Unfortunately I don't know of a way of making a windows app talk to a Firefox extension...

I may be completely wrong here, but some extensions can talk with installed apps, such as download manager extensions sending downloads to the managers, so maybe you could use whatever approach they use only in reverse. Just a thought. I personally prefer apps for stuff like this, so I don't even know if this issue would be important to me, but I figured I'd throw the idea out there in case it's actually helpful.

**Comment by punassuming** [comment]
As a Linux user, what are the other options? I'm getting to the point where I have to have chromium installed just to use Join

**Comment by vertigo220** [comment]
Obviously not a solution to get Join working, but you could try KDE Connect.

**Comment by dthigpen** [comment]
> 
> 
> Obviously not a solution to get Join working, but you could try KDE Connect.

Just to update this, KDE Connect was just removed from Google Play, but it can still downloaded from F-Droid.

**Comment by punassuming** [comment]
I don't know electron that well, but is it possible to use that as a application for Join?
